File: backend/service/quiz_exam_service/app/schemas/quiz_submission.py
from pydantic import BaseModel, Field, EmailStr
from uuid import UUID
from typing import List, Optional
from app.models.enum import SubmissionStatus , QuestionType, QuizType
from datetime import datetime
from app.schemas.submission_detail import QuestionGradeInput

class QuizSubmissionCreate(BaseModel):
    quiz_id: UUID
    user_id: UUID
    attempt_number: Optional[int] = None
    is_peer_review: Optional[bool] = False

class QuizSubmissionUpdate(BaseModel):
    status: SubmissionStatus | None = None
    submitted_at: Optional[datetime] = None
    # Chấm điểm chéo
    peer_avg_score: Optional[float] = None
    is_discrepant: bool | None = None
    completed_review_count: int | None = None 
    
    total_score: Optional[float] = None 
    is_passed: Optional[bool] = None 
    grader_id: Optional[UUID] = None

class SubmissionStatusOption(BaseModel):
    option_id: UUID
    option_text: str
    is_correct: Optional[bool] = None  # Chỉ có giá trị khi bài đã được chấm

class SubmissionStatusDetail(BaseModel):
    detail_id: UUID
    question_id: UUID
    question_title: str
    question_type: QuestionType
    video_trigger_seconds: Optional[int] = None
    body_content: Optional[str]
    max_points: Optional[float]
    options: List[SubmissionStatusOption]

    selected_option_id: Optional[UUID] = None
    essay_answer_text: Optional[str] = None
    graph_json_data: Optional[str] = None
    graph_image_url: Optional[str] = None

    # Chỉ trả khi bài đã được chấm
    score_earned: Optional[float] = None
    teacher_feedback: Optional[str] = None

class QuizSubmissionStatusResponse(BaseModel):
    submission_id: UUID
    quiz_id: UUID
    status: SubmissionStatus
    attempt_number: int
    started_at: datetime
    total_score: Optional[float] = None
    is_passed: Optional[bool] = None
    is_peer_review: Optional[bool] = None
    questions: List[SubmissionStatusDetail]


class QuizSubmissionSummaryResponse(BaseModel):
    """Schema trả về danh sách Quiz và trạng thái tổng quan các lượt nộp"""
    quiz_id: UUID = Field(..., description="Mã bài thi")
    subject_id: UUID = Field(..., description="Mã môn học")
    title: str = Field(..., description="Tiêu đề bài thi")
    description: Optional[str] = Field(None, description="Mô tả bài thi")
    duration_minutes: int = Field(..., description="Thời lượng làm bài (phút)")
    quiz_type: QuizType = Field(..., description="Loại đề thi (FIXED_QUESTION hoặc RANDOM_QUESTION)")
    is_active: bool = Field(..., description="Trạng thái bài thi")
    
    total_submissions: int = Field(default=0, description="Tổng số lượt làm bài")
    pending_gradings: int = Field(default=0, description="Số lượt chờ chấm (status = SUBMITTED)")
    graded_count: int = Field(default=0, description="Số lượt đã hoàn tất chấm (status = GRADED)")

class SubmissionListItemResponse(BaseModel):
    """Schema thông tin một lượt nộp bài hiển thị trên danh sách bảng"""
    submission_id: UUID = Field(..., description="Mã lượt nộp bài")
    quiz_id: UUID = Field(..., description="Mã bài thi")
    user_id: UUID = Field(..., description="Mã học viên")
    user_name: str = Field(..., description="Họ tên học viên (lấy từ User Service)")
    user_email: EmailStr = Field(..., description="Email học viên (lấy từ User Service)")
    
    attempt_number: int = Field(..., description="Lần làm bài thứ mấy")
    status: SubmissionStatus = Field(..., description="Trạng thái lượt nộp (IN_PROGRESS, SUBMITTED, GRADED)")
    started_at: datetime = Field(..., description="Thời điểm bắt đầu")
    submitted_at: Optional[datetime] = Field(None, description="Thời điểm nộp bài")
    
    total_score: Optional[float] = Field(None, description="Tổng điểm lượt làm bài")
    is_passed: Optional[bool] = Field(None, description="Đã đạt bài thi hay chưa")

class GradeSubmissionRequest(BaseModel):
    """Request Body khi giảng viên bấm 'Lưu kết quả chấm điểm'"""
    grader_id: UUID = Field(..., description="Mã giảng viên thực hiện chấm bài")
    grades: List[QuestionGradeInput] = Field(..., min_items=1, description="Danh sách điểm chấm từng câu")


class GradeSubmissionResponse(BaseModel):
    """Response trả về sau khi cập nhật điểm thành công"""
    submission_id: UUID = Field(..., description="Mã lượt nộp bài")
    status: SubmissionStatus = Field(..., description="Trạng thái sau chấm (Thường là GRADED)")
    total_score: float = Field(..., description="Tổng điểm sau khi tính toán")
    is_passed: bool = Field(..., description="Kết quả Đạt / Không đạt")
    grader_id: UUID = Field(..., description="Mã giảng viên chấm")
    graded_at: datetime = Field(..., description="Thời điểm hoàn tất chấm điểm")

class QuizUserSummaryResponse(BaseModel):
    user_id: UUID
    username: str
    email: EmailStr
    total_submissions: int
    pending_gradings: int
    latest_submitted_at: Optional[datetime] = None

class UserSubmissionItem(BaseModel):
    submission_id: UUID
    quiz_id: UUID
    user_id: UUID
    total_score: Optional[float] = None
    max_score: Optional[float] = None
    status: SubmissionStatus
    started_at: datetime
    submitted_at: Optional[datetime] = None

class QuestionGradingInput(BaseModel):
    detail_id: UUID
    score_earned: float
    teacher_feedback: Optional[str] = None

class GradeSubmissionRequest(BaseModel):
    gradings: List[QuestionGradingInput]

class QuizSubmissionStatus(BaseModel):
    submit_status: Optional[str] = None
    is_peer_review: bool | None = None

File: frontend/src/types/quiz-submission.ts
import { SubmissionStatus } from "./statuses";
import { QuizType } from "./exam-management";
export enum QuestionType {
    MULTIPLE_CHOICE = "MULTIPLE_CHOICE",
    ESSAY = "ESSAY",
    TRUE_FALSE = "TRUE_FALSE",
    FILL_IN_BLANK = "FILL_IN_BLANK",
}

export interface QuestionOption {
    option_id: string;
    option_text: string;
}

export interface QuizQuestion {
    detail_id: string;
    question_id: string;
    question_title: string;
    question_type: QuestionType;
    body_content: string | null;
    max_points: number;
    options: QuestionOption[];
    // 🆕 Mốc giây trong video mà câu hỏi này được kích hoạt (null nếu quiz không phải IN_VIDEO)
    video_trigger_seconds?: number | null;
    // 🆕 Dùng khi resume submission IN_PROGRESS đang dang dở (vd load lại trang giữa video)
    selected_option_id?: string | null;
    // 🆕 true/false nếu câu đã được chấm, null nếu chưa trả lời hoặc là ESSAY
    is_answered_correct?: boolean | null;
}

export interface QuizTakeResponse {
    submission_id: string;
    quiz_id: string;
    title: string;
    quiz_type: string;
    attempt_number: number;
    questions: QuizQuestion[];
}

export interface SubmissionDetailUpdatePayload {
    selected_option_id?: string | null;
    essay_answer_text?: string | null;
}

export interface SubmissionDetailUpdateResponse {
    message: string;
    detail_id: string;
    selected_option_id: string | null;
    essay_answer_text: string | null;
}

export interface QuizSubmitResponse {
    message: string;
    submission_id: string;
    status: string;
    is_passed: boolean;
    total_score: number;
    next_lesson_unlocked: boolean;
}

export interface SubmissionStatusOption {
    option_id: string;
    option_text: string;
    is_correct?: boolean | null; // Chỉ có khi status = GRADED
}

export interface SubmissionStatusDetail {
    detail_id: string;
    question_id: string;
    question_title: string;
    question_type: QuestionType;
    video_trigger_seconds?: number | null;
    body_content?: string | null;
    max_points?: number | null;
    options: SubmissionStatusOption[];
    selected_option_id?: string | null;
    essay_answer_text?: string | null;
    graph_json_data?: string | null;
    graph_image_url?: string | null;
    score_earned?: number | null;     // Chỉ có khi status = GRADED
    teacher_feedback?: string | null; // Chỉ có khi status = GRADED
}

export interface QuizSubmissionStatusResponse {
    submission_id: string;
    quiz_id: string;
    title?: string | null;
    quiz_type?: string | null;
    status: SubmissionStatus;
    attempt_number: number;
    started_at: string;
    total_score?: number | null;
    is_peer_review?: boolean | null;
    is_passed?: boolean | null;
    questions: SubmissionStatusDetail[];
}

export interface QuizStatusActionResult {
    success: boolean;
    data?: QuizSubmissionStatusResponse | null; // null nghĩa là chưa từng làm bài
    error?: string;
}

export interface QuizSubmissionSummary {
    quiz_id: string;
    subject_id: string;
    title: string;
    description: string | null;
    duration_minutes: number;
    quiz_type: QuizType;
    is_active: boolean;
    total_submissions: number;
    pending_gradings: number;
    graded_count: number;
}

export interface QuizUserSummaryItem {
    user_id: string;
    username: string;
    email: string;
    total_submissions: number;
    pending_gradings: number;
    latest_submitted_at: string | null;
}

export interface UserSubmissionItem {
    submission_id: string;
    quiz_id: string;
    user_id: string;
    total_score: number | null;
    max_score: number | null
    status: 'IN_PROGRESS' | 'SUBMITTED' | 'GRADED';
    started_at: string;
    submitted_at: string | null;
}

export interface QuestionOptionDetail {
    option_id: string;
    option_text: string;
    is_correct: boolean;
}

export interface SubmissionAnswerDetail {
    detail_id: string;
    question_id: string;
    question_text: string;
    body_content?: string | null; // 🆕 Nội dung câu có chỗ trống (dùng cho FILL_IN_BLANK)
    question_type: string;
    max_score: number;
    score_earned: number | null;
    selected_option_id: string | null;
    essay_answer_text: string | null;
    graph_json_data: string | null;
    graph_image_url: string | null;
    teacher_feedback: string | null;
    options: QuestionOptionDetail[];
}

export interface QuizSubmissionDetail {
    submission_id: string;
    quiz_id: string;
    quiz_title: string;
    user_id: string;
    username: string;
    email: string;
    score: number | null;
    max_possible_score: number;
    status: 'IN_PROGRESS' | 'SUBMITTED' | 'GRADED';
    started_at: string;
    submitted_at: string | null;
    answers: SubmissionAnswerDetail[];
}

export interface QuestionGradingPayload {
    detail_id: string;
    score_earned: number;
    teacher_feedback?: string;
}